File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/day3/AcWing1221/main.go
package main
/**
 * @Author 袁佳林
 * @Description unhappy
 * @Date 2023/2/20 11:22
 */
import (
    "bufio"
    "fmt"
    "os"
)

var (
    in = bufio.NewReader(os.Stdin)
    N  = 5000005
)

func main() {
    C := make([]int, N)
    D := make([]int, N)
    var n int
    _, err := fmt.Fscan(in, &n)
    if err != nil {
        return
    }
    for i := range C {
        C[i] = -1
    }
    for i := 0; i*i <= n; i++ {
        for j := i; j*j+i*i <= n; j++ {
            v := i*i + j*j
            if C[v] == -1 {
                C[v] = i
                D[v] = j
            }
        }
    }
    for i := 0; i*i <= n; i++ {
        for j := i; j*j+i*i <= n; j++ {
            v := n - i*i - j*j
            if C[v] != -1 {
                fmt.Println(i, j, C[v], D[v])
                return
            }
        }
    }
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/day3/AcWing1221/main.py
# author: 袁佳林
n = int(input())
c = [-1 for i in range(n + 10)]
d = [-1 for i in range(n + 10)]
m = int(n ** (1 / 2)) + 1
for i in range(m):
    for j in range(i, m):
        a = i ** 2 + j ** 2
        if a <= n and c[a] == -1:
            c[a] = i
            d[a] = j
for i in range(m):
    for j in range(i, m):
        b = n - i ** 2 - j ** 2
        if b >= 0 and ~c[b]:
            print(i, j, c[b], d[b])
            exit()

